h2/src/main/org/h2/util/ToDateTokenizer.java

```diff
@@ -0,0 +1,548 @@
+            patternToUse = Pattern.compile(format("^(%s)", name()), Pattern.CASE_INSENSITIVE);
+        }
+
+        private static List<FormatTokenEnum> EMPTY_LIST = new ArrayList<FormatTokenEnum>(0);
+
+        /**
+         * OPTIMISATION: Only return a list of {@link FormatTokenEnum} that share the same 1st char
+         * using the 1st char of the 'to parse' formatStr. Or return empty list if no match.
+         */
+        static List<FormatTokenEnum> getTokensInQuestion(final ToDateParams params) {
+            List<FormatTokenEnum> result = EMPTY_LIST;
+            if (cache.size() <= 0) {
+                initCache();
+            }
+            String formatStr = params.getFormatStr();
+            if (formatStr != null && formatStr.length() > 0) {
+                Character key = Character.toUpperCase(formatStr.charAt(0));
+                result = cache.get(key);
+            }
+            if (result == null) {
+                result = EMPTY_LIST;
+            }
+            return result;
+        }
+
+        private static synchronized void initCache() {
+            if (cache.size() <= 0) {
+                for (FormatTokenEnum token : FormatTokenEnum.values()) {
+
+                    List<Character> tokenKeys = new ArrayList<Character>();
+
+                    if(token.name().contains("_")) {
+                       String[] tokens = token.name().split("_");
+                       for(String tokenLets : tokens) {
+                           tokenKeys.add(tokenLets.toUpperCase().charAt(0));
+                       }
+                    } else {
+                       tokenKeys.add(token.name().toUpperCase().charAt(0));
+                    }
+
+                    for(Character tokenKey : tokenKeys) {
+                        List<FormatTokenEnum> l = cache.get(tokenKey);
+                        if (l == null) {
+                            l = new ArrayList<FormatTokenEnum>(1);
+                            cache.put(tokenKey, l);
+                        }
+                        l.add(token);
+                    }
+                }
+            }
+
+        }
+
+        /**
+         * Parse the format-string with passed token of {@link FormatTokenEnum}}.<br>
+         * if token matches return true otherwise false.
+         */
+        boolean parseFormatStrWithToken(final ToDateParams params) {
+            Matcher matcher = patternToUse.matcher(params.getFormatStr());
+            boolean foundToken = matcher.find();
+            if (foundToken) {
+                String formatTokenStr = matcher.group(1);
+                toDateParslet.parse(params, this, formatTokenStr);
+            }
+            return foundToken;
+        }
+    }
+
+    /**
+     * Interface of the classes that can parse a specialized small bit of the TO_DATE format-string
+     */
+    interface ToDateParslet {
+        ToDateParams parse(ToDateParams params, FormatTokenEnum formatTokenEnum, String formatTokenStr);
+    }
```

Comment:
all of the implementations of this method seem to return the params parameter, which makes the return value superfluous.